verifier/src/x509.rs

```diff
@@ -6,5 +6,177 @@ mod certs;
+};
+
+#[derive(Clone)]
+pub struct TrustAnchor(MbedtlsList<Certificate>);
+
+/// Try to get a trust anchor from a PEM-encoded string.
+///
+/// # Errors
+/// `Error::MbedTls` if the string is not valid PEM certificate.
+impl TryFrom<&str> for TrustAnchor {
+    type Error = Error;
+
+    fn try_from(pem: &str) -> Result<Self> {
+        Self::try_from(String::from(pem))
+    }
+}
+
+/// Try to get a trust anchor from a PEM-encoded string.
+///
+/// # Errors
+/// `Error::MbedTls` if the string is not valid PEM certificate.
+impl TryFrom<String> for TrustAnchor {
+    type Error = Error;
+
+    fn try_from(mut pem: String) -> Result<Self> {
+        pem.push('\0');
+        let certs = Certificate::from_pem_multiple(pem.as_bytes())?;
+        Ok(Self(certs))
+    }
+}
+
+/// An unverified certificate chain.
+///
+/// This is mostly opaque meant to be used to verify and create a
+/// [`VerifiedCertChain`].
+#[derive(Clone)]
+pub struct UnverifiedCertChain(MbedtlsList<Certificate>);
+
+impl UnverifiedCertChain {
+    /// Verify the certificate chain is valid for the given `trust_anchor`.
+    ///
+    /// # Errors
+    /// `Error::MbedTls` if the certificate chain is not valid.
+    pub fn verify(self, trust_anchor: &TrustAnchor) -> Result<VerifiedCertChain> {
+        let profile = Profile::new(
+            vec![HashType::Sha256, HashType::Sha384, HashType::Sha512],
+            // The note on `PkType::Ecdsa` is a lie:
+            //
+            // > This type is never returned by the mbedTLS key parsing routines
+            //
+            // It comes back when using the Intel cert chain.
+            vec![PkType::Rsa, PkType::Eckey, PkType::Ecdsa],
+            vec![
+                EcGroupId::Curve25519,
+                EcGroupId::SecP256K1,
+                EcGroupId::SecP256R1,
+                EcGroupId::SecP384R1,
+                EcGroupId::SecP521R1,
+            ],
+            2048,
+        );
+        Certificate::verify_with_profile(&self.0, &trust_anchor.0, None, Some(&profile), None)?;
+        Ok(VerifiedCertChain(self.0))
+    }
+}
+
+/// Try to get a certificate chain from a PEM-encoded string.
+///
+/// # Errors
+/// `Error::MbedTls` if the string is not valid PEM certificate(s).
+impl TryFrom<&str> for UnverifiedCertChain {
+    type Error = Error;
+
+    fn try_from(pem: &str) -> Result<Self> {
+        Self::try_from(String::from(pem))
+    }
+}
+
+/// Try to get a certificate chain from a PEM-encoded string.
+///
+/// # Errors
+/// `Error::MbedTls` if the string is not valid PEM certificate(s).
+impl TryFrom<String> for UnverifiedCertChain {
+    type Error = Error;
+
+    fn try_from(mut pem: String) -> Result<Self> {
+        pem.push('\0');
+        let certs = Certificate::from_pem_multiple(pem.as_bytes())?;
+        Ok(Self(certs))
+    }
+}
+
+/// A verified certificate chain.
+///
+/// See [`UnverifiedCertChain::verify`] for creating one.
+pub struct VerifiedCertChain(MbedtlsList<Certificate>);
```

Comment:
I'm still on the fence on this design of Unverified -> Verified.
I did it to nudge a client into verifying before doing anything else with the certchain

Reply:
It seems like a good idea to me